stdlib/camlinternalLazy.ml

```diff
@@ -59,83 +54,73 @@ external reset_to_lazy : Obj.t -> unit = "caml_lazy_reset_to_lazy" [@@noalloc]
+  Obj.set_field b 0 (Obj.repr ()); (* Release the closure *)
   let result = closure () in
   Obj.set_field b 0 (Obj.repr result);
   update_to_forward b;
   result
 
+(* Called by [force_gen] *)
 let force_gen_lazy_block ~only_val (blk : 'arg lazy_t) =
+  (* In the common case, expect the tag to be [lazy_tag], but may be other tags
+     due to concurrent forcing of lazy values. *)
   match update_to_forcing (Obj.repr blk) with
   | 0 when only_val -> do_force_val_block blk
   | 0 -> do_force_block blk
-  | 1 -> raise Undefined
   | _ ->
-      (* We raise [RacyLazy] for all the other cases.
-       *
-       * If the return value was 2, we should raise RacyLazy. We had observed
-       * earlier in [force_gen] that the [blk] was not [Obj.forward_tag] and
-       * not short-circuited. But now we observe that the value has either been
-       * forwarded (return value 3) or forwarded and short-circuited (4). Both
-       * of these indicate a race. *)
-      raise RacyLazy
+      (* If the return value is 1, then the tag is [forcing_tag]. We are either
+         recursively forcing on the same domain, or concurrently forcing from
+         multiple domains. Raise [Undefined].
+
+         If the return value is 2, then the tag is [forward_tag]. In
+         [force_gen], we checked whether the tag was [forward_tag]. It has
+         changed since the last read. Hence, there is another domain which is
+         concurrently forcing this lazy. Raise [Undefined].
+
+         If the return value is 3, then the tag is not [forcing_tag] or
+         [forward_tag], which we checked in [force_gen]. The tag is not
+         [lazy_tag] or otherwise [update_to_forcing] would have succeeded and
+         returned [0]. The tag has changed from the expected [lazy_tag] due to
+         another domain concurrently, and the GC short circuiting the result.
+         Raise [Undefined].
+         *)
+      raise Undefined
 
 (* used in the %lazy_force primitive *)
 let force_lazy_block blk = force_gen_lazy_block ~only_val:false blk
 
-let try_force_gen_lazy_block ~only_val (blk : 'arg lazy_t) =
-  match update_to_forcing (Obj.repr blk) with
-  | 0 when only_val -> Some (do_force_val_block blk)
-  | 0 -> Some (do_force_block blk)
-  | 1 -> raise Undefined
-  | _ ->
-      (* Other cases indicate racy access to lazy for the same reasons outlined
-       * in [force_gen_lazy_block]. *)
-      None
-
 (* [force_gen ~only_val:false] is not used, since [Lazy.force] is
    declared as a primitive whose code inlines the tag tests of its
    argument, except when afl instrumentation is turned on. *)
-
 let force_gen ~only_val (lzv : 'arg lazy_t) =
   let x = Obj.repr lzv in
-  (* XXX No safe points start *)
   let t = Obj.tag x in
-  if t = Obj.forward_tag then (Obj.obj (Obj.field x 0) : 'arg)
-  (* XXX No safe points end *)
-  else if t <> Obj.lazy_tag && t <> Obj.forcing_tag then (Obj.obj x : 'arg)
+  (* START no safe points. If a GC occurs here, then the object [x] may be
```

Comment:
I'll move it to where it was. [@poll error] is not sufficient as it cannot reason about calls to functions IINM. For example, The analysis would encounter the function call `Obj.obj` and report that it is not safe. Is that correct @sadiqj?

Reply:
Moved.

Reply:
Trying on trunk (similar code), the poll-error attribute does not see Obj.tag or Obj.obj as polling, but it does see Obj.field as polling (whether it actually polls or not on trunk, there is something to do about it).

You still need a notation to allow a call to a polling function as suggested here: https://github.com/ocaml/ocaml/pull/10462#issuecomment-868457097 (namely the call to `force_gen_lazy_block`) because there is no obvious way to refactor it (apart maybe by conveying the need to evaluate the thunk using an exception?).